<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Обязательная-часть" data-toc-modified-id="Обязательная-часть-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Обязательная часть</a></span><ul class="toc-item"><li><span><a href="#Задание-1.-Базовое-изучение" data-toc-modified-id="Задание-1.-Базовое-изучение-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Задание 1. Базовое изучение</a></span></li><li><span><a href="#Задание-2.-Работа-с-выбросами" data-toc-modified-id="Задание-2.-Работа-с-выбросами-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Задание 2. Работа с выбросами</a></span></li><li><span><a href="#Задание-3.-Работа-с-пропусками" data-toc-modified-id="Задание-3.-Работа-с-пропусками-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Задание 3. Работа с пропусками</a></span></li></ul></li></ul></div>

## Обязательная часть

Будем осуществлять работу с непростым [набором данных](https://raw.githubusercontent.com/obulygin/pyda_homeworks/master/statistics_basics/horse_data.csv) о состоянии здоровья лошадей, испытывающих кишечные колики. 

### Задание 1. Базовое изучение

Изучить представленный набор данных на основе [описания его столбцов](https://raw.githubusercontent.com/obulygin/pyda_homeworks/master/statistics_basics/horse_data.names) и выбрать 8 столбцов для дальнейшего изучения (среди них должны быть как числовые, так и категориальные). Провести расчет базовых метрик для них, кратко описать результаты.

In [1]:
from bs4 import BeautifulSoup
import pandas as pd
import requests
import numpy as np
pd.set_option('display.max_rows', 500)

In [84]:
horses = pd.read_csv('horse_data.csv', na_values = "?")

horses = horses.rename(columns={'2': 'Операции', '1': 'Возраст', '530101': '№ больничный', '38.50': 't ректальная', 
                        '66': 'Пульс', '28': 'частота дыхания', '3': 't конечностей', 
                        '3.1': 'Периферический пульс', '?': 'Слизистая', '2.1': 'время наполнения капилляров',
                        '5': 'Болевые ощущения', '4': 'Перистальтика', '4.1': 'Вздутие живота',
                        '?.1': 'назогастральный зонд', '?.2': 'назогастральный рефлюкс', '?.3': 'назогастральный рефлюкс PH',
                        '3.2': 'анализ кала', '5.1': 'Брюшная полость', '45.00': 'гематокрит', '8.40': 'Всего белка',
                        '?.4': 'внешний вид пункции', '?.5': 'Всего белка в пункции', '2.2': 'Итог лечения',
                        '2.3': 'Хирургические осложнения', '11300': 'тип осложнений', '00000': 'type of lesion-1',
                        '00000.1': 'type of lesion 2', '2.4': 'cp_data'})

horses.info()
# horses.head()
# horses.describe()
# horses[horses['Пульс'] == 30.0]
# horses[horses['Возраст'] == 9]
# horses[horses['Всего белка'] == 89.000000]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 299 entries, 0 to 298
Data columns (total 28 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   Операции                     298 non-null    float64
 1   Возраст                      299 non-null    int64  
 2   № больничный                 299 non-null    int64  
 3   t ректальная                 239 non-null    float64
 4   Пульс                        275 non-null    float64
 5   частота дыхания              241 non-null    float64
 6   t конечностей                243 non-null    float64
 7   Периферический пульс         230 non-null    float64
 8   Слизистая                    253 non-null    float64
 9   время наполнения капилляров  267 non-null    float64
 10  Болевые ощущения             244 non-null    float64
 11  Перистальтика                255 non-null    float64
 12  Вздутие живота               243 non-null    float64
 13  назогастральный зонд

**Решение**

Было принято решение взять следующие колонки:

***Операции*** -  один из самых заполненных показателей (298), определяющий наличие операции у лошади (категориальный).

***Возраст***  - один из самых заполненных показателей (299), определяющий возраст лошади (категориальный).

***Пульс***  - один из самых заполненных показателей (275), определяющий пульс лошади (числовой).

***Перистальтика*** - достаточно хорошо заполненный показатель (255), важный, наряду с вздутием (категориальный).

***Вздутие живота*** - достаточно хорошо заполненный показатель (255), важный, для определения причин колик (категориальный).

***Всего белка*** - достаточно хорошо заполненный показатель (255), важный, для определения обезвоживания (числовой).

***Итог лечения*** - один из самых заполненных показателей (298), определяющий исход лечения: выжила лошадь, или нет.

***Периферический пульс*** - не самый заполненный показатель (233), но может помочь найти в будущем замену NaN в пульсе

<!-- 0   операции*                    298 non-null    float64 - актуально общий показатель* (категориальное)
1   возраст*                     299 non-null    int64  - актуально* (категориальное) так-то числовое, но здесь категориальное
2   № больничный                 299 non-null    int64  - неактуально
3   t ректальная                 239 non-null    float64 - актуально 
4   Пульс                        275 non-null    float64 - актуально (большое кол-во данных)* (числовое)
5   частота дыхания              241 non-null    float64 - относительно много данных
6   t конечностей                243 non-null    float64 - относительно много данных
7   Периферический пульс         230 non-null    float64 - относительно много данных
8   Слизистая                    253 non-null    float64 - относительно много данных
9   время наполнения капилляров  267 non-null    float64 - относительно много данных, но мало относится к предмету
10  болевые ощущения*            244 non-null    float64 - акутально * (категориальное)
11  перистальтика                255 non-null    float64 - акутально 
12  вздутие живота*              243 non-null    float64 - актуально * (категориальное)
13  назогастральный зонд         196 non-null    float64 - газы из зонда: неактуально, поскольку естьболее полные данные по вздутию живота
14  назогастральный рефлюкс      194 non-null    float64 - а нужен ли нам анализ рвоты?
15  назогастральный рефлюкс PH   53 non-null     float64 - слишком мало данных по кислотности рвоты 
16  анализ кала                  197 non-null    float64 - а нужен ли нам анализ кала?
17  Брюшная полость              181 non-null    float64 - актуально вид брюшной полости 
18  гематокрит                   270 non-null    float64 - актуально - осадок эритроцитов 
19  Всего белка                  266 non-null    float64 - акутально* (числовое)
20  внешний вид пункции          135 non-null    float64 - мало актуально в силу менее 50% данных
21  Всего белка в пункции        102 non-null    float64 - мало,поскольку пункцию проводили всего 135 раз 
22  Решение*                     298 non-null    float64 - актуально * (категориальное)
23  хирургические осложнения     299 non-null    int64  - актуально * (категориальное)
24  type of lesion               299 non-null    int64  - неактуально, поскольку не знаем специфику
25  type of lesion-1             299 non-null    int64  - 0 значения  - не актуально
26  type of lesion 2             299 non-null    int64  - 0 значения  - не актуально
27  cp_data                      299 non-null    int64  - эта переменная не имеет значения, поскольку данные патологии не включается и не собирается для этих случаев  -->

**Начинаем работать с данными**

1. получаем базовые метрики horses_data:

In [85]:
horses_data = horses[['Операции', 'Возраст', 'Пульс', 'Перистальтика', 'Вздутие живота', 'Всего белка', 'Итог лечения', 'Периферический пульс']]
# horses_data.info()
horses_data.describe()
# horses_data.head()

Операции    Возраст       Пульс  Перистальтика  Вздутие живота  \
count  298.000000  299.00000  275.000000     255.000000      243.000000   
mean     1.395973    1.64214   71.934545       2.913725        2.259259   
std      0.489881    2.17730   28.680522       0.976298        1.061472   
min      1.000000    1.00000   30.000000       1.000000        1.000000   
25%      1.000000    1.00000   48.000000       3.000000        1.000000   
50%      1.000000    1.00000   64.000000       3.000000        2.000000   
75%      2.000000    1.00000   88.000000       4.000000        3.000000   
max      2.000000    9.00000  184.000000       4.000000        4.000000   

       Всего белка  Итог лечения  Периферический пульс  
count   266.000000    298.000000            230.000000  
mean     24.517293      1.550336              2.013043  
std      27.509056      0.737967              1.042672  
min       3.300000      1.000000              1.000000  
25%       6.500000      1.000000              1.000000  
50%       7.500000      1.000000              1.500000  
75%      57.000000      2.000000              3.000000  
max      89.000000      3.000000              4.000000

**Заменяем NaN на 0**

Сначала заменяем NaN на 0, поскольку это позволит производить подсчеты в циклах, а также избавит от некоторых проблем (например, при работе с выбросами, если в Data Frame есть NAN то при выделении выбросов, после удаления диапазона межквартильного размаха в выборку попадают значения не с выбросами, а с NaN).

В текущем DataFrame такая замена позволительна, поскольку в итоге все нулевые значения также будут заменены или в крайнем случае удалены.

In [111]:
horses_no_null = horses_data.fillna(0)
horses_no_null.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 299 entries, 0 to 298
Data columns (total 8 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Операции              299 non-null    float64
 1   Возраст               299 non-null    int64  
 2   Пульс                 299 non-null    float64
 3   Перистальтика         299 non-null    float64
 4   Вздутие живота        299 non-null    float64
 5   Всего белка           299 non-null    float64
 6   Итог лечения          299 non-null    float64
 7   Периферический пульс  299 non-null    float64
dtypes: float64(7), int64(1)
memory usage: 18.8 KB


### Задание 2. Работа с выбросами

В выбранных числовых столбцах найти выбросы, выдвинуть гипотезы об их причинах и проинтерпретировать результаты. Принять и обосновать решение о дальнейшей работе с ними.

**Решение**

Исходя из данных вышеприведенной сводной таблицы выброы могут присутствовать в колонках:
- Пульс
- всего белка

поскольку данные колонки содержатколичественные, а не категориальные данные.

Исключением является колонка **Возраст**, в которой при возможных всего двух параметрах 1 и 2 существуют 24 строки с параметром 9, что явно является ошибкой. Поскольку значений равных 2 в колонке "Возраст" нет вообще (хотя, должны быть, поскольку по ним есть описание), можно сделать вывод, что 9 и есть на самом деле 2. Таким образом, просто заменим значение 9 в колонке "Возраст" на значение 2. 

In [112]:
def true_age(row):
    if row['Возраст'] == 9: return 2
    else: return row['Возраст']
    return row['Возраст']

horses_no_null['Возраст'] = horses_no_null.apply(true_age, axis=1)
horses_no_null.head()

Операции  Возраст  Пульс  Перистальтика  Вздутие живота  Всего белка  \
0       1.0      1.0   88.0            4.0             2.0         85.0   
1       2.0      1.0   40.0            3.0             1.0          6.7   
2       1.0      2.0  164.0            4.0             4.0          7.2   
3       2.0      1.0  104.0            0.0             0.0          7.4   
4       2.0      1.0    0.0            3.0             2.0          0.0   

   Итог лечения  Периферический пульс  
0           3.0                   0.0  
1           1.0                   1.0  
2           2.0                   1.0  
3           2.0                   0.0  
4           1.0                   1.0

**Вернемся к выбросам:**

***Определяем выбросы в колонке "Пульс":***

In [113]:
hq1 = horses_no_null['Пульс'].quantile(0.25)
hq3 = horses_no_null['Пульс'].quantile(0.75)
iqr = hq3 - hq1
lower_bound = hq1 - (1.5 * iqr) 
upper_bound = hq3 + (1.5 * iqr)
horses_no_outliers_by_puls = horses_no_null[horses_no_null['Пульс'].between(lower_bound, upper_bound, inclusive=True)]
horses_no_outliers_by_puls.describe()
horses_no_outliers_by_puls.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 294 entries, 0 to 298
Data columns (total 8 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Операции              294 non-null    float64
 1   Возраст               294 non-null    float64
 2   Пульс                 294 non-null    float64
 3   Перистальтика         294 non-null    float64
 4   Вздутие живота        294 non-null    float64
 5   Всего белка           294 non-null    float64
 6   Итог лечения          294 non-null    float64
 7   Периферический пульс  294 non-null    float64
dtypes: float64(8)
memory usage: 20.7 KB


***Просто смотрим, сколько у нас выбросов по Пульсу:***


In [114]:
horses_pulls_outliers = pd.concat([horses_no_null, horses_no_outliers_by_puls]).drop_duplicates(keep=False)
horses_pulls_outliers.count()

Операции                5
Возраст                 5
Пульс                   5
Перистальтика           5
Вздутие живота          5
Всего белка             5
Итог лечения            5
Периферический пульс    5
dtype: int64

***Определяем выбросы в колонке "Всего белка":***

In [115]:
hpq1 = horses_no_null['Пульс'].quantile(0.25)
hpq3 = horses_no_null['Пульс'].quantile(0.75)
piqr = hq3 - hq1
lower_bound = hpq1 - (1.5 * piqr) 
upper_bound = hpq3 + (1.5 * piqr)
horses_no_outliers_by_protein = horses_no_null[horses_no_null['Всего белка'].between(lower_bound, upper_bound, inclusive=True)]
horses_no_outliers_by_protein.describe()
horses_no_outliers_by_protein.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 299 entries, 0 to 298
Data columns (total 8 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Операции              299 non-null    float64
 1   Возраст               299 non-null    float64
 2   Пульс                 299 non-null    float64
 3   Перистальтика         299 non-null    float64
 4   Вздутие живота        299 non-null    float64
 5   Всего белка           299 non-null    float64
 6   Итог лечения          299 non-null    float64
 7   Периферический пульс  299 non-null    float64
dtypes: float64(8)
memory usage: 21.0 KB


Как видно из результов (в dataframe, отфильтрованный помежквартильному размаху попали все 299 строк), ***выбросы в колонке "Всего белка" отсутствуют.***

Для дальнейшей работы будет использован файл, в котором убраны выбросы по колонке "Пульс". Переназначим его обратно в ***horses_no_null***

In [116]:
horses_no_null = horses_no_outliers_by_puls
horses_no_null.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 294 entries, 0 to 298
Data columns (total 8 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Операции              294 non-null    float64
 1   Возраст               294 non-null    float64
 2   Пульс                 294 non-null    float64
 3   Перистальтика         294 non-null    float64
 4   Вздутие живота        294 non-null    float64
 5   Всего белка           294 non-null    float64
 6   Итог лечения          294 non-null    float64
 7   Периферический пульс  294 non-null    float64
dtypes: float64(8)
memory usage: 20.7 KB


### Задание 3. Работа с пропусками

Рассчитать количество выбросов для всех выбранных столбцов. Принять и обосновать решение о методе работы с пропусками по каждому столбцу, сформировать датафрейм, в котором пропуски будут отсутствовать.

**3. Работаем с заполнением пропусков**

Исходя из данных horses_data.info() мы понимаем, что в текущем варианте условно слабым местом у нас являются колонки:

- Пульс: 275 значений из 299 (не самый плохой вариант)
- Перистальтика: 255 из 299 (низкое количество)
- Вздутие живота: 243 из 299 (самое низкое количество)
- Всего белка: 266 из 299 (достаточно низкое количество)

Исходя из специфики текущего массива данных, оптимальным видится не удаление пропусков, а их заполнение медианными значениями, поскольку даже в выбранных 8 столбцах можно проследить общие закономерности и взаимосвязи. Таким образом, задача сводится к:

- выявлению взаимосвязей
- заполнению NaN (в нашем случае замененных на 0) медианными значениями на основе выявленных взаимосвязей (паттернов)

In [117]:
horses_no_null[horses_no_null['Пульс'] == 0].head(30)

Операции  Возраст  Пульс  Перистальтика  Вздутие живота  Всего белка  \
4         2.0      1.0    0.0            3.0             2.0          0.0   
27        1.0      1.0    0.0            0.0             0.0          0.0   
51        2.0      1.0    0.0            3.0             1.0          7.7   
55        1.0      1.0    0.0            0.0             0.0          6.7   
57        1.0      1.0    0.0            4.0             3.0          5.9   
73        1.0      2.0    0.0            0.0             0.0          4.9   
77        1.0      1.0    0.0            3.0             3.0          5.9   
82        1.0      1.0    0.0            0.0             4.0          7.8   
92        2.0      1.0    0.0            3.0             3.0          6.5   
114       2.0      1.0    0.0            3.0             2.0         70.0   
116       2.0      1.0    0.0            4.0             3.0          6.7   
125       1.0      1.0    0.0            4.0             2.0          6.5   
135       1.0      1.0    0.0            3.0             3.0          5.9   
150       1.0      1.0    0.0            4.0             2.0          0.0   
158       2.0      1.0    0.0            1.0             2.0          7.5   
159       1.0      1.0    0.0            0.0             0.0         58.0   
173       2.0      1.0    0.0            0.0             4.0          0.0   
174       1.0      1.0    0.0            0.0             0.0          0.0   
181       1.0      1.0    0.0            0.0             4.0          7.8   
203       1.0      1.0    0.0            4.0             2.0          6.6   
206       1.0      1.0    0.0            0.0             0.0          8.7   
215       2.0      1.0    0.0            0.0             0.0          0.0   
226       1.0      2.0    0.0            0.0             0.0          6.5   
287       1.0      1.0    0.0            0.0             0.0          0.0   

     Итог лечения  Периферический пульс  
4             1.0                   1.0  
27            2.0                   0.0  
51            1.0                   1.0  
55            1.0                   0.0  
57            2.0                   3.0  
73            2.0                   0.0  
77            2.0                   3.0  
82            2.0                   3.0  
92            2.0                   3.0  
114           1.0                   1.0  
116           3.0                   3.0  
125           1.0                   1.0  
135           2.0                   3.0  
150           2.0                   3.0  
158           1.0                   1.0  
159           1.0                   0.0  
173           2.0                   4.0  
174           1.0                   0.0  
181           2.0                   3.0  
203           1.0                   1.0  
206           2.0                   0.0  
215           1.0                   0.0  
226           1.0                   0.0  
287           1.0                   0.0

**Вышеприведенная фильтрация:**
- демонстрирует результат по самому заполненному из интересуемых нас показателю, а также наличиюзначений в других столбцах.
- следовательно, дает нам возможность заполнить сначала максимально столбец "пульс" (например, исходя из взаимосвязи с показателем периферического пульса, но не только), а потом уже возвращаться к остальным столбцам.

**белок/пульс:**

(распределение взято на основе квантилей)

In [132]:
print('При показателе белка от 3,3 до 6,5:')
print('медиана по пульсу:', horses_data[(horses_data['Всего белка'] >= 3.3)&(horses_data['Всего белка'] < 6.5)]['Пульс'].median())
print('среднее по пульсу:', horses_data[(horses_data['Всего белка'] >= 3.3)&(horses_data['Всего белка'] < 6.5)]['Пульс'].mean(),'\n')
print('При показателе белка от 6.5 до 7.5:')
print('медиана по пульсу:', horses_data[(horses_data['Всего белка'] >= 6.5)&(horses_data['Всего белка'] < 7.5)]['Пульс'].median())
print('среднее по пульсу:', horses_data[(horses_data['Всего белка'] >= 6.5)&(horses_data['Всего белка'] < 7.5)]['Пульс'].mean(),'\n')
print('При показателе белка от 7.5 до 57:')
print('медиана по пульсу:', horses_data[(horses_data['Всего белка'] >= 7.5)&(horses_data['Всего белка'] < 57)]['Пульс'].median())
print('среднее по пульсу:', horses_data[(horses_data['Всего белка'] >= 7.5)&(horses_data['Всего белка'] < 57)]['Пульс'].mean(),'\n')
print('При показателе белка от 57 до 89:')
print('медиана по пульсу:', horses_data[(horses_data['Всего белка'] >= 57)&(horses_data['Всего белка'] < 89)]['Пульс'].median())
print('среднее по пульсу:', horses_data[(horses_data['Всего белка'] >= 57)&(horses_data['Всего белка'] < 89)]['Пульс'].mean(),'\n')


При показателе белка от 3,3 до 6,5:
медиана по пульсу: 82.0
среднее по пульсу: 80.36 

При показателе белка от 6.5 до 7.5:
медиана по пульсу: 55.0
среднее по пульсу: 63.294117647058826 

При показателе белка от 7.5 до 57:
медиана по пульсу: 70.0
среднее по пульсу: 77.0 

При показателе белка от 57 до 89:
медиана по пульсу: 60.0
среднее по пульсу: 66.27692307692308 



Данные вполне вписываются в общую картину, поскольку сказано, что чем выше белок, тем больше обезвоживание. Обратный тезис хорошо виден на значениях пониженного белка (до 6,5), говорящий о высоком содержании воды в организме и логичном высоком пульсе (много воды - повышается внутренне давление - учащается пульс).

При этом относительно завышенные показатели медианы при высоком белке (при белке от 57 медиана попульсу 60, хотя норма 30-40) могут быть связаны с другими факторами (мы, все же больных лошадей исследуем, а не здоровых)

**Взаимосвязь между пульсом и периферическим пульсом:**

In [129]:
print('Периферический пульс = 1:')
print('Пульс, медиана:', horses_data[(horses_data['Периферический пульс'] == 1)]['Пульс'].median())
print('Пульс, среднее:', horses_data[(horses_data['Периферический пульс'] == 1)]['Пульс'].mean(),'\n')
print('Периферический пульс = 2:')
print('Пульс, медиана:', horses_data[(horses_data['Периферический пульс'] == 2)]['Пульс'].median())
print('Пульс, среднее:', horses_data[(horses_data['Периферический пульс'] == 2)]['Пульс'].mean(),'\n')
print('Периферический пульс = 3:')
print('Пульс, медиана:', horses_data[(horses_data['Периферический пульс'] == 3)]['Пульс'].median())
print('Пульс, среднее:', horses_data[(horses_data['Периферический пульс'] == 3)]['Пульс'].mean(),'\n')
print('Периферический пульс = 4:')
print('Пульс, медиана:', horses_data[(horses_data['Периферический пульс'] == 4)]['Пульс'].median())
print('Пульс, среднее:', horses_data[(horses_data['Периферический пульс'] == 4)]['Пульс'].mean())

Периферический пульс = 1:
Пульс, медиана: 52.0
Пульс, среднее: 58.0 

Периферический пульс = 2:
Пульс, медиана: 60.0
Пульс, среднее: 59.2 

Периферический пульс = 3:
Пульс, медиана: 88.0
Пульс, среднее: 87.79787234042553 

Периферический пульс = 4:
Пульс, медиана: 88.0
Пульс, среднее: 91.42857142857143


Поскольку медианы по пульсу по 3 и 4 категории совпадают (что также не есть хорошо и в конечном итоге может слишком сгладить результаты), по 4 категории возьмем значение среднего арифметического, т.е., 91,4

***Т.о., для заполнения максимально возможного количества пустых ячеек в столбце "Пульс" пишем функцию:***

In [118]:
def fill_na_puls_from_protein_and_pp(row):
    """
    функция находит значения в колонке 'Пульс' на основании медианных значений при схожих 
    показателях в колонках "Всего белка" и "Периферический пульс"
    """
    if row['Пульс'] == 0:
        if row['Всего белка'] != 0:
            if row['Всего белка'] >= 3.3 and row['Всего белка'] <6.5: return 82.0
            elif row['Всего белка'] >= 6.3 and row['Всего белка'] <7.5: return 55.0
            elif row['Всего белка'] >= 7.5 and row['Всего белка'] <57: return 70.0 
            elif row['Всего белка'] >= 57.0 and row['Всего белка'] <89: return 60.0 
            else: return row['Пульс']
        else:
            if row['Периферический пульс'] == 1.0: return 52.0
            elif row['Периферический пульс'] == 2.0: return 60.0
            elif row['Периферический пульс'] == 3.0: return 88.0
            elif row['Периферический пульс'] == 4.0: return 91.4
            else: return row['Пульс']
    else: return row['Пульс']
    
horses_no_null['Пульс'] = horses_no_null.apply(fill_na_puls_from_protein_and_pp, axis=1)
horses_no_null[horses_no['Пульс'] == 0].head(50)
horses_no_null.head(5)

Операции  Возраст  Пульс  Перистальтика  Вздутие живота  Всего белка  \
0       1.0      1.0   88.0            4.0             2.0         85.0   
1       2.0      1.0   40.0            3.0             1.0          6.7   
3       2.0      1.0  104.0            0.0             0.0          7.4   
4       2.0      1.0   52.0            3.0             2.0          0.0   
5       1.0      1.0   48.0            3.0             3.0          7.0   

   Итог лечения  Периферический пульс  
0           3.0                   0.0  
1           1.0                   1.0  
3           2.0                   0.0  
4           1.0                   1.0  
5           1.0                   1.0

**Идем дальше**

**Взаимосвязь между пульсом и перистальтикой**

Использование диапазонов квантилей для получения медианных значений дало очень плохие результаты

In [71]:
print('от 0 до 1 квантиля:', horses_data[(horses_data['Пульс'] >= 30)&(horses_data['Пульс'] < 48)]['Перистальтика'].median())
print('от 1 до 2 квантиля:', horses_data[(horses_data['Пульс'] >= 48)&(horses_data['Пульс'] < 64)]['Перистальтика'].median())
print('от 2 до 3 квантиля:', horses_data[(horses_data['Пульс'] >= 64)&(horses_data['Пульс'] < 88)]['Перистальтика'].median())
print('от 3 квантиля:', horses_data[(horses_data['Пульс'] >88)]['Перистальтика'].median())


от 0 до 1 квантиля: 3.0
от 1 до 2 квантиля: 3.0
от 2 до 3 квантиля: 3.0
от 3 квантиля: 4.0


**Однако опытным путем удалось выделить несколько разных групп:**

In [ ]:
horses_data[(horses_data['Пульс'] == 50)]['Перистальтика'].median()

***Чем ниже пульс, тем выраженнее перистальтика***, но, исходя из сравнений медианных значений:

диапазон от 30 до 48 - перистальтика = 3

диапазон от 48 до 51 - перистальтика = 2 -2,5

диапазон от 51 до 85 - перистальтика = 3

диапазон от 86 - перистальтика = 4

In [121]:
def fill_na_peristalsis_by_puls(row):
    if row['Перистальтика'] == 0:
        if row['Пульс'] >=48 and row['Пульс'] <=51: return 2.0
        elif row['Пульс'] >= 30 and row['Пульс'] <48 or row['Пульс'] > 51 and row['Пульс'] <=85: return 3.0
        elif row['Пульс'] >=86.0: return 4.0 
        else: return 0
    else: return row['Перистальтика']
    
horses_no_null['Перистальтика'] = horses_no_null.apply(fill_na_peristalsis_by_puls, axis=1)
horses_no_null[horses_no_null['Перистальтика'] == 0].head(50)
horses_no_null.head(5)

Операции  Возраст  Пульс  Перистальтика  Вздутие живота  Всего белка  \
0       1.0      1.0   88.0            4.0             2.0         85.0   
1       2.0      1.0   40.0            3.0             1.0          6.7   
3       2.0      1.0  104.0            4.0             0.0          7.4   
4       2.0      1.0   52.0            3.0             2.0          0.0   
5       1.0      1.0   48.0            3.0             3.0          7.0   

   Итог лечения  Периферический пульс  
0           3.0                   0.0  
1           1.0                   1.0  
3           2.0                   0.0  
4           1.0                   1.0  
5           1.0                   1.0

**Взаимосвязь между перистальтикой и вздутием** 

**Чем  больше вздутие, тем меньше перистальтика.** 

Это ясно из описания столбцов (тупо читаем документацию). 
Собственно, это же и подтверждается расчетом мединных значений по категориям, относщимся к сильной и крайне сильной степени вздутия живота (3, 4). Однако, самое веселое начинается, когдамы начинаем тестировать категории 2 и 1 (у вздутия это "отсутствует" и "легкое", а у перистальтики это "сниженная" и "нормальная" соответственно).

При любых значениях вздутия (говорим про 1, 2) перистальтика показывает 3 категорию: 

In [58]:
print('При вздутии живота категории 1 медиана по перистальтики: ', horses_data[(horses_data['Вздутие живота'] == 1)]['Перистальтика'].median())
print('При вздутии живота категории 2 медиана по перистальтики: ', horses_data[(horses_data['Вздутие живота'] == 2)]['Перистальтика'].median())
print('При вздутии живота категории 3 медиана по перистальтики: ', horses_data[(horses_data['Вздутие живота'] == 3)]['Перистальтика'].median())
print('При вздутии живота категории 4 медиана по перистальтики: ', horses_data[(horses_data['Вздутие живота'] == 4)]['Перистальтика'].median())

При вздутии живота категории 1 медиана по перистальтики:  3.0
При вздутии живота категории 2 медиана по перистальтики:  3.0
При вздутии живота категории 3 медиана по перистальтики:  3.0
При вздутии живота категории 4 медиана по перистальтики:  4.0


***Однако, все меняется, когда мы делаем проверку вздутия по перистальтике***

(никогда так много не писал про вздутие и перистальтику):

In [60]:
print('При перистальтике категории 1 медиана по вздутию: ', horses_data[(horses_data['Перистальтика'] == 1)]['Вздутие живота'].median())
print('При перистальтике категории 2 медиана по вздутию: ', horses_data[(horses_data['Перистальтика'] == 2)]['Вздутие живота'].median())
print('При перистальтике категории 3 медиана по вздутию: ', horses_data[(horses_data['Перистальтика'] == 3)]['Вздутие живота'].median())
print('При перистальтике категории 4 медиана по вздутию: ', horses_data[(horses_data['Перистальтика'] == 4)]['Вздутие живота'].median())

При перистальтике категории 1 медиана по вздутию:  1.0
При перистальтике категории 2 медиана по вздутию:  1.0
При перистальтике категории 3 медиана по вздутию:  2.0
При перистальтике категории 4 медиана по вздутию:  3.0


**Что же делать, как же быть?**

Исходя из логики, для малых значений (1, 2) правильнее будет взять за основу медианы вздутия по Перистальтике, а не наоборот.

Во-первых, ясно сказано, что при малых значниях вздутия перистальтика с высокой долей вероятности будет более выражена, что явно не попадает в категорию 3.

Во-вторых, данных по перистальтике несколько больше, чем по вздутию (255 против 243), что дает нам условно большую уверенностьв правильности за счет несколько большей выборки


In [122]:
def fill_na_gaz(row):
    if row['Вздутие живота'] == 0:
        if row['Перистальтика'] == 4.0: return 3.0
        elif row['Перистальтика'] <=2.0: return  1.0
        elif row['Перистальтика'] ==3.0: return 2.0 
        else: return row['Вздутие живота']
    else: return row['Вздутие живота']
    
horses_no_null['Вздутие живота'] = horses_no_null.apply(fill_na_gaz, axis=1)
# horses_no_null.head(300)
horses_no_null[horses_no_null['Пульс'] == 0].head(150)


Операции  Возраст  Пульс  Перистальтика  Вздутие живота  Всего белка  \
27        1.0      1.0    0.0            0.0             1.0          0.0   
174       1.0      1.0    0.0            0.0             1.0          0.0   
215       2.0      1.0    0.0            0.0             1.0          0.0   
287       1.0      1.0    0.0            0.0             1.0          0.0   

     Итог лечения  Периферический пульс  
27            2.0                   0.0  
174           1.0                   0.0  
215           1.0                   0.0  
287           1.0                   0.0

***Теперь самым незаполненными остаются столбцы "Всего белка" и "Периферический пульс", поскольку с ними мы так  не работали.***

Для обоих этих столбцов предлагается использоват функции, обратные по смыслу тем, что заполняли столбец "Пульс" и провести замену 0 значений медианами на основе попадания значений в столбце пульс в тот или иной диапазон.

Безусловно, данный подходимеет массу минусов, поскольку усредняет итоговые значения, но выбор невелик и это все равно лучше, чем оставлять ли удалять пропуски

Диапазанное значения можно брать двух видов:
    1. на основе квантилей (вслучае с взаимоотношением "Пульс/Перистальтика" такой подход не очень себя зарекомендовал, но тут все может быть иначе).
    2. на основе собственных выборок (см. "Пульс/Перистальтика")

**результаты по диапазонам найденным опытным путем:**

In [78]:
print('в диапазоне от 30 до 48:', horses_data[(horses_data['Пульс'] >=30)&(horses_data['Пульс'] <=48)]['Всего белка'].median())
print('в диапазоне от 48 до 51:', horses_data[(horses_data['Пульс'] >=48)&(horses_data['Пульс'] <=51)]['Всего белка'].median())
print('в диапазоне от 51 до 85:', horses_data[(horses_data['Пульс'] >51)&(horses_data['Пульс'] <=85)]['Всего белка'].median())
print('в диапазоне от 86:', horses_data[(horses_data['Пульс'] >=86)]['Всего белка'].median())

в диапазоне от 30 до 48: 7.55
в диапазоне от 48 до 51: 7.3
в диапазоне от 51 до 85: 7.5
в диапазоне от 86: 7.7


**результаты по диапазонам на основе квантилей:**

In [75]:
print('от 0 до 1 квантиля:', horses_data[(horses_data['Пульс'] >= 30)&(horses_data['Пульс'] < 48)]['Всего белка'].median())
print('от 1 до 2 квантиля:', horses_data[(horses_data['Пульс'] >= 48)&(horses_data['Пульс'] < 64)]['Всего белка'].median())
print('от 2 до 3 квантиля:', horses_data[(horses_data['Пульс'] >= 64)&(horses_data['Пульс'] < 88)]['Всего белка'].median())
print('от 3 квантиля:', horses_data[(horses_data['Пульс'] >88)]['Всего белка'].median())


от 0 до 1 квантиля: 7.65
от 1 до 2 квантиля: 7.5
от 2 до 3 квантиля: 7.4
от 3 квантиля: 8.0


В данном случае медианы на основе диапазонов по квантилям имеют больший разброс, а следовательно теоретически больше нам подходят (в теории, зависимость между пульсом и давлением, а значит и уровнем дегидратации/белка должна быть более явной).
Их будем использовать.

In [123]:

def fill_na_protein_from_puls(row):
    """
    функция помогает найти значение в колонке 'Всего белка' на основании данных колонки 'Пульс'
    """
    if row['Всего белка'] == 0:
        if row['Пульс'] >= 30 and row['Пульс'] < 48: return 7.65
        elif row['Пульс'] >= 48 and row['Пульс'] < 64: return 7.5
        elif row['Пульс'] >= 64 and row['Пульс'] < 88: return 7.4
        elif row['Пульс'] >= 88: return 8.0
        else: return row['Пульс']
    else: return row['Всего белка']
    
horses_no_null['Всего белка'] = horses_no_null.apply(fill_na_protein_from_puls, axis=1)
horses_no_null[horses_no_null['Всего белка'] == 0.0].head(50)
# horses_no_null.head(300)

Операции  Возраст  Пульс  Перистальтика  Вздутие живота  Всего белка  \
27        1.0      1.0    0.0            0.0             1.0          0.0   
174       1.0      1.0    0.0            0.0             1.0          0.0   
215       2.0      1.0    0.0            0.0             1.0          0.0   
287       1.0      1.0    0.0            0.0             1.0          0.0   

     Итог лечения  Периферический пульс  
27            2.0                   0.0  
174           1.0                   0.0  
215           1.0                   0.0  
287           1.0                   0.0

***Теперь разбираемся с Периферическим пульсом.***

Здесь все проще, поскольку зависимость между пульсом и периферическим пульсом гораздо выраженнее. Используем зависимость, обратную зависимости "Перферический пульс"/"Пульс"

У нас есть следующие данные:

Периферический пульс = 1:
Пульс, медиана: 52.0

Периферический пульс = 2:
Пульс, медиана: 60.0

Периферический пульс = 3:
Пульс, медиана: 88.0

Периферический пульс = 4:
Пульс, среднее: 91.4

в таком случае можно разбить на следующие диапазоны:
- если пульс в пределах до 56 (56 = 52 + ((60-52)/2)), то периферический пульс = 1
- если пульс в пределах от 57 до 74 (74 = 60 + ((88-60)/2), то периферический пульс = 2
- если пульс в пределах от 75 до 89,7 (89,7 = 75 + ((91,4-88)/2), то периферический пульс = 3
- если пульс в пределах от 89,8 и выше, то периферический пульс = 4

In [124]:
def fill_na_pp_from_puls(row):
    """
    функция помогает найти значение в колонке 'Периферический пульс' на основании данных колонки 'Пульс'
    """
    if row['Периферический пульс'] == 0:
        if row['Пульс'] != 0:
            if row['Пульс'] <= 56: return 1.0
            elif row['Пульс'] >= 57 and row['Пульс'] <= 74: return 2.0
            elif row['Пульс'] >= 75 and row['Пульс'] < 89.7: return 3.0
            elif row['Пульс'] >= 89.8: return 4.0
            else: return row['Периферический пульс']
        else: return row['Периферический пульс']
    else: return row['Периферический пульс']
    
horses_no_null['Периферический пульс'] = horses_no_null.apply(fill_na_pp_from_puls, axis=1)
horses_no_null[horses_no_null['Периферический пульс'] == 0.0].head(50)
# horses_no_null.head(300)

Операции  Возраст  Пульс  Перистальтика  Вздутие живота  Всего белка  \
27        1.0      1.0    0.0            0.0             1.0          0.0   
174       1.0      1.0    0.0            0.0             1.0          0.0   
215       2.0      1.0    0.0            0.0             1.0          0.0   
287       1.0      1.0    0.0            0.0             1.0          0.0   

     Итог лечения  Периферический пульс  
27            2.0                   0.0  
174           1.0                   0.0  
215           1.0                   0.0  
287           1.0                   0.0

Таким образом, на текущий момент у нас есть всего 5 строк, в которых по какой-либо колонке присутствуют значения = 0.
При этом:
- в 4 из 5 строк значения отсутствуют (равны 0) сразу в 4 колонках из 8
- в 1 строке значения отсутствуют в 2 колонках из 8 и оба этих значения исключительно категориальные, информирующие о наличии операции и итоге лечения.

Произвести корректную замену в подобном случае достаточно сложно, поскольку требуется искать более сложные и неявные взимосвязи между значениями столбцов (уверен, в теории это возможно, если ты ветеринар 85 lvl).

5 строк - это 1,67 % от всех данных, что достаточно мало для влияния на метрики, а следовательно этими строками можно пренебречь, просто удалив их.

In [125]:
horses_all_filled = horses_no_null.drop(horses_no_null[(horses_no_null['Пульс'] == 0) & (horses_no_null['Операции'] == 0 )].index)
horses_all_filled.head(5)

Операции  Возраст  Пульс  Перистальтика  Вздутие живота  Всего белка  \
0       1.0      1.0   88.0            4.0             2.0         85.0   
1       2.0      1.0   40.0            3.0             1.0          6.7   
3       2.0      1.0  104.0            4.0             3.0          7.4   
4       2.0      1.0   52.0            3.0             2.0          7.5   
5       1.0      1.0   48.0            3.0             3.0          7.0   

   Итог лечения  Периферический пульс  
0           3.0                   3.0  
1           1.0                   1.0  
3           2.0                   4.0  
4           1.0                   1.0  
5           1.0                   1.0

In [126]:
# horses_data.info()
horses_data.describe()

Операции    Возраст       Пульс  Перистальтика  Вздутие живота  \
count  298.000000  299.00000  275.000000     255.000000      243.000000   
mean     1.395973    1.64214   71.934545       2.913725        2.259259   
std      0.489881    2.17730   28.680522       0.976298        1.061472   
min      1.000000    1.00000   30.000000       1.000000        1.000000   
25%      1.000000    1.00000   48.000000       3.000000        1.000000   
50%      1.000000    1.00000   64.000000       3.000000        2.000000   
75%      2.000000    1.00000   88.000000       4.000000        3.000000   
max      2.000000    9.00000  184.000000       4.000000        4.000000   

       Всего белка  Итог лечения  Периферический пульс  
count   266.000000    298.000000            230.000000  
mean     24.517293      1.550336              2.013043  
std      27.509056      0.737967              1.042672  
min       3.300000      1.000000              1.000000  
25%       6.500000      1.000000              1.000000  
50%       7.500000      1.000000              1.500000  
75%      57.000000      2.000000              3.000000  
max      89.000000      3.000000              4.000000

In [127]:
horses_all_filled.describe()

Операции     Возраст       Пульс  Перистальтика  Вздутие живота  \
count  294.000000  294.000000  294.000000     294.000000      294.000000   
mean     1.394558    1.064626   69.161224       2.891156        2.190476   
std      0.496511    0.246284   26.535257       0.981948        1.000569   
min      0.000000    1.000000    0.000000       0.000000        1.000000   
25%      1.000000    1.000000   48.000000       3.000000        1.000000   
50%      1.000000    1.000000   62.000000       3.000000        2.000000   
75%      2.000000    1.000000   88.000000       4.000000        3.000000   
max      2.000000    2.000000  146.000000       4.000000        4.000000   

       Всего белка  Итог лечения  Периферический пульс  
count   294.000000    294.000000            294.000000  
mean     22.831633      1.540816              1.993197  
std      26.686725      0.745888              1.070846  
min       0.000000      0.000000              0.000000  
25%       6.600000      1.000000              1.000000  
50%       7.500000      1.000000              2.000000  
75%      54.000000      2.000000              3.000000  
max      89.000000      3.000000              4.000000

**Итог замены.**

Метрики практически не изменились. Самый большой сдвиг произошел в колонке "Всего белка":
- в стандартизированном отклонении (итоговое снизилось почти на 1 единицу (0,822331)),
- в 75 процентиле (итоговое снизилось на 3 единицы).